pkg/kn/commands/plugin/path_verifier.go

```diff
@@ -0,0 +1,102 @@
+// Copyright © 2018 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+// PathVerifier receives a path and determines if it is valid or not
+type PathVerifier interface {
+	// Verify determines if a given path is valid
+	Verify(path string) []error
+}
+
+// CommandOverrideVerifier verifies that existing kn commands are not overriden
+type CommandOverrideVerifier struct {
+	Root        *cobra.Command
+	SeenPlugins map[string]string
+}
+
+// Verify implements PathVerifier and determines if a given path
+// is valid depending on whether or not it overwrites an existing
+// kn command path, or a previously seen plugin.
+func (v *CommandOverrideVerifier) Verify(path string) []error {
+	if v.Root == nil {
+		return []error{fmt.Errorf("unable to verify path with nil root")}
+	}
+
+	// extract the plugin binary name
+	segs := strings.Split(path, "/")
+	binName := segs[len(segs)-1]
+
+	cmdPath := strings.Split(binName, "-")
```

Comment:
Yup, you can and it becomes: `kn list types` where list is the command group.

Try it! :)

Reply:
I *want* to be able to make `kn list-types`

Reply:
Afaiu the `kubectl` logic (and also the implementation here when executing the plugin), it's that for a plugin `kn-service-list_types` you can call `kn service list-types` (so you can't have commands with underscores in it though). This is not reflected in the in this validation code here though when checking for shadowing existing commands).

I think, when using `kn plugin list` this file -> command translation should be shown there as well, otherwise when only the plain file name is shown, the user would need to know about this convention to find out how to call the plugin.

Reply:
```bash
➜  client git:(kn-plugins4) cat /usr/local/bin/kn-foo_bar
#!/bin/bash

echo "Hello Knative, I'm a Kn plugin"
➜  client git:(kn-plugins4) cat /usr/local/bin/kn-foo-bar
#!/bin/bash

echo "Hello Knative, I'm a Kn plugin"
➜  client git:(kn-plugins4) ./kn --lookup-plugins-in-path=true plugin list
Using kn config file: /Users/maximilien/.kn/config.yaml
The following compatible plugins are available, using options:
  - plugins dir: '~/.kn/plugins'
  - lookup plugins in path: 'true'

/Users/maximilien/.kn/plugins/kn-hello
/usr/local/bin/kn-foo-bar
/usr/local/bin/kn-foo_bar
/usr/local/bin/kn-hello
  - warning: /usr/local/bin/kn-hello is overshadowed by a similarly named plugin: /Users/maximilien/.kn/plugins/kn-hello
/usr/local/bin/kn-kn
/usr/local/bin/kn-knctl

error: one plugin warning was found

➜  client git:(kn-plugins4) ./kn --lookup-plugins-in-path=true foo-bar
Hello Knative, I'm a Kn plugin
➜  client git:(kn-plugins4) ./kn --lookup-plugins-in-path=true foo bar
Hello Knative, I'm a Kn plugin
```